packages/gen-ai-hub/src/utils/deployment-resolver.ts

```diff
@@ -27,56 +65,121 @@ export interface FoundationModel {
+/**
+ * Query the AI Core service for a deployment that matches the given criteria. If more than one deployment matches the criteria, the first one is returned.
+ * @param opts - The options for the deployment resolution.
+ * @returns A promise of a deployment, if a deployment was found, fails otherwise.
+ */
+export async function resolveDeployment(
+  opts: DeploymentResolutionOptions
+): Promise<AiDeployment> {
+  const { model } = opts;
 
-  if (opts.model) {
-    const modelName = opts.model.name;
-    deploymentList = deploymentList.filter(
-      deployment => extractModel(deployment)?.name === modelName
+  let deployments = await getAllDeployments(opts);
+
+  if (model) {
+    deployments = deployments.filter(
+      deployment => extractModel(deployment)?.name === model.name
     );
-    if (opts.model.version) {
-      const modelVersion = opts.model.version;
-      // feature idea: smart handling of 'latest' version: treat 'latest' and the highest version number as the same
-      deploymentList = deploymentList.filter(
-        deployment => extractModel(deployment)?.version === modelVersion
+
+    if (model.version) {
+      deployments = deployments.filter(
+        deployment => extractModel(deployment)?.version === model.version
       );
     }
   }
 
-  if (!deploymentList.length) {
+  if (!deployments.length) {
     throw new Error(
       'No deployment matched the given criteria: ' + JSON.stringify(opts)
     );
   }
-  return deploymentList[0];
+  return deployments[0];
+}
+
+async function getAllDeployments(
+  opts: DeploymentResolutionOptions
+): Promise<AiDeployment[]> {
+  const { scenarioId, executableId, groupId = 'default' } = opts;
+  // TODO: add a cache: https://github.tools.sap/AI/gen-ai-hub-sdk-js-backlog/issues/78
+  try {
+    return (
+      await DeploymentApi.deploymentQuery(
+        {
+          scenarioId,
+          status: 'RUNNING',
+          ...(executableId && { executableIds: [executableId] })
+        },
+        { 'AI-Resource-Group': groupId }
+      ).execute()
+    ).resources;
+  } catch (error) {
+    throw new Error('Failed to fetch the list of deployments: ' + error);
+  }
+}
+
+function extractModel(
+  deployment: AiDeployment
+): Partial<FoundationModel> | undefined {
+  return deployment.details?.resources?.backend_details?.model;
+}
+
+/**
+ * Get the deployment ID for a given model deployment configuration and executable ID.
+ * @param modelDeployment - The model deployment configuration.
+ * @param executableId - The executable ID.
+ * @param requestConfig - The request configuration.
+ * @returns The ID of the deployment, if found.
+ */
+export async function getDeploymentId(
+  modelDeployment: ModelDeployment,
+  executableId: string,
+  requestConfig?: CustomRequestConfig
+): Promise<string> {
+  if (isDeploymentIdConfiguration(modelDeployment)) {
+    return modelDeployment.deploymentId;
+  }
+
+  return (
+    await resolveDeployment({
+      scenarioId: 'foundation-models',
```

Comment:
[pp] This function is for a specific scenario-id, maybe add this to the description?